File: scaling-controller/solace_autoscale/assignment/topology.py
"""Shard topology and the ordering-first key-to-broker model (event spine, ADR 0009).

This module is **pure**: no I/O, no clock, no logging. It defines the self-describing topology
snapshot the control plane publishes on the event spine, the deterministic way a partition key maps
to a broker within a shard, and the minimal set of handoffs a topology change produces.

Why it lives here and stays pure:
  - The spine publisher (the only component that touches the bus) turns a ``ShardTopology`` into the
    JSON event; the Go shim parses the same JSON and computes the same owner for a key. A committed
    golden vector proves the two agree, exactly like the rule spec.
  - Ordering is the priority. Keys map to brokers by **rendezvous (highest-random-weight) hashing**
    so a scale from N to N+1 brokers moves only about 1/(N+1) of the keys instead of reshuffling all
    of them. Fewer moved keys means fewer per-key handoffs, which means less opportunity to reorder a
    key's stream. ``diff`` reports exactly which keys' owners changed as ``Handoff`` records, each
    fenced by the generation at which the cutover becomes effective.

The generation (``gen``) is monotonic per shard. It is both the dedup/version for idempotent event
application (a consumer ignores any snapshot with ``gen`` less than or equal to the one it applied)
and the fence for ordering (a message cut over to a new broker is stamped with the gen so the
listener can hold it until the old broker's stream for that key has drained).
"""

from __future__ import annotations

import hashlib
from dataclasses import dataclass, field

from .store import Broker, BrokerState

#: Wire property name carrying the generation a message was published under. Stamped alongside the
#: partition key so a listener can preserve per-key order across a broker handoff.
GENERATION_PROPERTY = "saas_gen"

#: Topology event schema version. Bump when the event shape changes in a way consumers must gate on.
TOPOLOGY_VERSION = 1

#: A broker takes NEW keyed traffic only in this state (mirrors _ASSIGNABLE_STATES in the store).
_OWNABLE_STATES = {BrokerState.ACTIVE}


@dataclass(frozen=True)
class BrokerRef:
    """A broker as it appears in a topology snapshot: id, state, and its endpoint map (locations,
    never credentials)."""

    broker_id: str
    state: BrokerState
    endpoints: dict[str, str] = field(default_factory=dict)

    @classmethod
    def from_broker(cls, b: Broker) -> BrokerRef:
        return cls(broker_id=b.broker_id, state=b.state, endpoints=dict(b.endpoints))


@dataclass(frozen=True)
class Handoff:
    """One key-ownership move produced by a topology change. The shim publishes the moved key to
    ``from_broker`` until it is quiesced, then cuts it to ``to_broker`` stamped with
    ``effective_gen`` so the listener keeps the key's order across the cutover."""

    from_broker: str
    to_broker: str
    effective_gen: int
    reason: str = "topology-change"


@dataclass(frozen=True)
class ShardTopology:
    """Self-describing snapshot of one shard's brokers at a given generation.

    ``handoffs`` is empty in steady state and populated only across a change so a shim knows which
    cutovers to run drain-before-cutover for. Applying a snapshot needs no callback: it carries
    everything a shim needs to route and to hand keys off safely.
    """

    shard: str
    gen: int
    brokers: tuple[BrokerRef, ...] = ()
    handoffs: tuple[Handoff, ...] = ()

    def ownable_brokers(self) -> list[BrokerRef]:
        """Brokers that may own a key (ACTIVE), in deterministic broker_id order."""
        return sorted((b for b in self.brokers if b.state in _OWNABLE_STATES),
                      key=lambda b: b.broker_id)

    def owner(self, key: str) -> str | None:
        """The broker that owns ``key`` in this topology, by rendezvous hashing over the ownable
        brokers. Returns None if the shard has no ownable broker. Deterministic and identical to the
        Go shim's Owner."""
        return _rendezvous_owner(key, [b.broker_id for b in self.ownable_brokers()])

    # ---- event round-trip (the wire form the spine publishes) ------------------------------------

    def to_event(self, *, emitted_at: str) -> dict:
        """Render the portable topology event. ``emitted_at`` is an ISO-8601 timestamp supplied by
        the caller (this module never reads a clock)."""
        return {
            "version": TOPOLOGY_VERSION,
            "shard": self.shard,
            "gen": self.gen,
            "emitted_at": emitted_at,
            "brokers": [
                {"broker_id": b.broker_id, "state": b.state.value, "endpoints": dict(b.endpoints)}
                for b in self.brokers
            ],
            "handoffs": [
                {"from_broker": h.from_broker, "to_broker": h.to_broker,
                 "effective_gen": h.effective_gen, "reason": h.reason}
                for h in self.handoffs
            ],
        }

    @classmethod
    def from_event(cls, event: dict) -> ShardTopology:
        version = event.get("version")
        if version != TOPOLOGY_VERSION:
            raise ValueError(f"unsupported topology version {version!r} (expected {TOPOLOGY_VERSION})")
        brokers = tuple(
            BrokerRef(
                broker_id=b["broker_id"],
                state=BrokerState(b["state"]),
                endpoints=dict(b.get("endpoints", {})),
            )
            for b in event.get("brokers", [])
        )
        handoffs = tuple(
            Handoff(
                from_broker=h["from_broker"],
                to_broker=h["to_broker"],
                effective_gen=int(h["effective_gen"]),
                reason=h.get("reason", "topology-change"),
            )
            for h in event.get("handoffs", [])
        )
        return cls(shard=event["shard"], gen=int(event["gen"]), brokers=brokers, handoffs=handoffs)


def _rendezvous_owner(key: str, broker_ids: list[str]) -> str | None:
    """Highest-random-weight (rendezvous) hashing: the owner is the broker with the greatest
    ``hash(broker_id + "\\x00" + key)``. Ties break on broker_id so the result is total and
    deterministic. Moving from N to N+1 brokers reassigns a key only if the new broker outscores the
    incumbent, so about 1/(N+1) of keys move and the rest stay put."""
    if not broker_ids:
        return None
    best_id: str | None = None
    best_score = b""
    for bid in broker_ids:
        score = hashlib.sha256(f"{bid}\x00{key}".encode()).digest()
        if best_id is None or score > best_score or (score == best_score and bid > best_id):
            best_id, best_score = bid, score
    return best_id


def diff(old: ShardTopology, new: ShardTopology, keys: list[str]) -> list[Handoff]:
    """Handoffs implied by moving from ``old`` to ``new`` for a known set of partition ``keys``.

    A handoff is emitted for every key whose owner changes, deduplicated to one per
    (from_broker, to_broker) pair since the cutover protocol is per broker pair, not per key. The
    cutover is fenced at ``new.gen``. Keys that do not move produce nothing, which is the common case
    under rendezvous hashing.
    """
    seen: set[tuple[str, str]] = set()
    handoffs: list[Handoff] = []
    for key in keys:
        src = old.owner(key)
        dst = new.owner(key)
        if src is None or dst is None or src == dst:
            continue
        pair = (src, dst)
        if pair in seen:
            continue
        seen.add(pair)
        handoffs.append(Handoff(from_broker=src, to_broker=dst, effective_gen=new.gen))
    return handoffs

File: shim/topology/topology.go
// Package topology is the Go side of the event spine (ADR 0009). It parses the self-describing
// topology snapshot the Python control plane publishes on _autoscale/shard/<shard>/topology and
// answers the one question the shim needs at publish time: which broker owns a partition key.
//
// Ownership uses rendezvous (highest-random-weight) hashing, computed identically to the Python
// control plane: for a key, the owner is the ACTIVE broker with the greatest sha256(broker_id +
// "\x00" + key), ties broken on the greater broker_id. This must match Python byte-for-byte, which
// is what TestTopologyInteropGolden proves against a Python-emitted golden file. Rendezvous makes
// scaling N->N+1 move only ~1/(N+1) of keys, which is what keeps handoffs (and the ordering risk
// they carry) minimal.
//
// This package is pure: it parses bytes and computes ownership. It does no I/O and holds no bus
// dependency. The spine subscriber (a later package) feeds snapshots in; the resolver applies them.
package topology

import (
	"bytes"
	"crypto/sha256"
	"encoding/json"
	"fmt"
	"sort"
)

// Version is the topology event schema version this package understands. A snapshot with any other
// version is rejected rather than misparsed.
const Version = 1

// GenerationProperty is the message application property carrying the generation stamp. It matches
// the Python GENERATION_PROPERTY and the dispatch layer's saas_gen, so a subscriber can fence on the
// generation without decoding the body.
const GenerationProperty = "saas_gen"

// State mirrors the Python BrokerState. Only Active brokers own keys.
type State string

const (
	StateActive   State = "active"
	StateDraining State = "draining"
	StateDrained  State = "drained"
	StateDeleting State = "deleting"
	StateGone     State = "gone"
)

// BrokerRef is one broker in a snapshot: its id, state, and per-protocol endpoint map.
type BrokerRef struct {
	BrokerID  string            `json:"broker_id"`
	State     State             `json:"state"`
	Endpoints map[string]string `json:"endpoints"`
}

// Handoff records that a key range moved from one broker to another, fenced at EffectiveGen. The
// publisher keeps sending a moved key to FromBroker until it drains, then cuts to ToBroker stamping
// EffectiveGen; the listener holds gen-EffectiveGen traffic for a key until older traffic drains.
type Handoff struct {
	FromBroker   string `json:"from_broker"`
	ToBroker     string `json:"to_broker"`
	EffectiveGen int    `json:"effective_gen"`
	Reason       string `json:"reason"`
}

// Topology is a parsed snapshot for one shard. It is the wire event, plus ownership computation.
type Topology struct {
	Version    int         `json:"version"`
	Shard      string      `json:"shard"`
	Generation int         `json:"gen"`
	EmittedAt  string      `json:"emitted_at"`
	Brokers    []BrokerRef `json:"brokers"`
	Handoffs   []Handoff   `json:"handoffs"`
}

// LoadTopology parses a topology snapshot from its wire JSON, rejecting an unknown version.
func LoadTopology(data []byte) (*Topology, error) {
	var t Topology
	dec := json.NewDecoder(bytes.NewReader(data))
	if err := dec.Decode(&t); err != nil {
		return nil, fmt.Errorf("decode topology event: %w", err)
	}
	if t.Version != Version {
		return nil, fmt.Errorf("unsupported topology version %d (expected %d)", t.Version, Version)
	}
	return &t, nil
}

// Gen returns the snapshot's generation.
func (t *Topology) Gen() int { return t.Generation }

// AllHandoffs returns the in-flight handoffs carried by this snapshot.
func (t *Topology) AllHandoffs() []Handoff { return t.Handoffs }

// ownableBrokers returns the ACTIVE broker ids, sorted, matching Python's ownable_brokers().
func (t *Topology) ownableBrokers() []string {
	ids := make([]string, 0, len(t.Brokers))
	for _, b := range t.Brokers {
		if b.State == StateActive {
			ids = append(ids, b.BrokerID)
		}
	}
	sort.Strings(ids)
	return ids
}

// Owner returns the broker id that owns the partition key, or "" (ok=false) when no broker is
// ownable. Rendezvous: greatest sha256(broker_id + 0x00 + key), ties broken on the greater
// broker_id. Identical to the Python ShardTopology.owner.
func (t *Topology) Owner(key string) (string, bool) {
	ids := t.ownableBrokers()
	if len(ids) == 0 {
		return "", false
	}
	var bestID string
	var bestDigest [sha256.Size]byte
	first := true
	for _, id := range ids {
		h := sha256.New()
		h.Write([]byte(id))
		h.Write([]byte{0})
		h.Write([]byte(key))
		var digest [sha256.Size]byte
		h.Sum(digest[:0])
		if first {
			bestID, bestDigest, first = id, digest, false
			continue
		}
		cmp := bytes.Compare(digest[:], bestDigest[:])
		if cmp > 0 || (cmp == 0 && id > bestID) {
			bestID, bestDigest = id, digest
		}
	}
	return bestID, true
}

// EndpointFor returns the connection URI for a broker id and protocol, or an error if the broker is
// not in the snapshot or has no endpoint for that protocol.
func (t *Topology) EndpointFor(brokerID, protocol string) (string, error) {
	for _, b := range t.Brokers {
		if b.BrokerID == brokerID {
			uri, ok := b.Endpoints[protocol]
			if !ok {
				return "", fmt.Errorf("broker %q has no endpoint for protocol %q", brokerID, protocol)
			}
			return uri, nil
		}
	}
	return "", fmt.Errorf("broker %q not in topology for shard %q", brokerID, t.Shard)
}
